src/Dom/Document.php

```diff
@@ -2044,4 +2049,67 @@ public function enforceCssMaxByteCount($maxByteCount = AMP::MAX_CSS_BYTE_COUNT)
     {
         $this->cssMaxByteCountEnforced = $maxByteCount;
     }
+
+    /**
+     * Add a http-equiv charset meta tag to the document's <head> node.
+     *
+     * This is needed to make the DOMDocument behave as it should in terms of encoding.
+     * See: http://php.net/manual/en/domdocument.loadhtml.php#78243.
+     *
+     * @param string $html HTML string to add the http-equiv charset to.
+     * @return string Adapted string of HTML.
+     */
+    private function addHttpEquivCharset($html)
+    {
+        $count = 0;
+
+        // We try first to detect an existing <head> node.
+        $html = preg_replace(
+            self::HTML_GET_HEAD_OPENING_TAG_PATTERN,
+            self::HTML_GET_HEAD_OPENING_TAG_REPLACEMENT,
+            $html,
+            1,
+            $count
+        );
+
+        // In case no <head> node was found, we try to prepend it together with the http-equiv to the <body> tag.
+        if ($count < 1) {
+            $html = preg_replace(
+                self::HTML_GET_BODY_OPENING_TAG_PATTERN,
+                self::HTML_GET_BODY_OPENING_TAG_REPLACEMENT,
+                $html,
+                1,
+                $count
+            );
+        }
+
+        // If no <body> was found either, we look for the <html> tag instead.
+        if ($count < 1) {
+            $html = preg_replace(
+                self::HTML_GET_HTML_OPENING_TAG_PATTERN,
+                self::HTML_GET_HTML_OPENING_TAG_REPLACEMENT,
+                $html,
+                1,
+                $count
+            );
+        }
+
+        // If no <html> was found either, we look for the <!doctype> tag instead.
+        if ($count < 1) {
+            $html = preg_replace(
+                self::HTML_GET_HTML_OPENING_TAG_PATTERN,
+                self::HTML_GET_HTML_OPENING_TAG_REPLACEMENT,
+                $html,
+                1,
+                $count
+            );
+        }
+
+        // Finally, we just prepend the head with the required http-equiv charset.
+        if ($count < 1) {
+            $html = '<head>' . self::HTTP_EQUIV_META_TAG . '</head>' . $html;
+        }
```

Comment:
All of these may be a bit excessive. 

Consider this markup:

```php
$dom = new DOMDocument();
$dom->loadHTML( '<meta id="injected-meta-charset" http-equiv="content-type" content="text/html; charset=utf-8"><p>¡Fragment!</p>' );
echo $dom->saveHTML();
```

The output is:

```html
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><head><meta id="injected-meta-charset" http-equiv="content-type" content="text/html; charset=utf-8"></head><body><p>¡Fragment!</p></body></html>
```

So the parser automatically puts the `meta` charset in a `head` that it constructs dynamically.

I think all that is needed is if there is no `<html>` in the document, then just prepend the `meta` tag and the parser can fix it up.

Reply:
The `Dom\Document` has different default behavior than `DOMDocument`. Amongst others, it uses `LIBXML_HTML_NODEFDTD`, which prevents the default doctype (and as a side-effect, the `<head>` node) to be added automatically, as that was needed to allow for comments before the doctype (see https://github.com/ampproject/amp-wp/issues/4104#issuecomment-574216012).
I'll have to test how adding a meta tag without a `<head>` node behaves now.

Reply:
Hmm, the HTML one was in there twice, and the doctype one was not actually implemented... 🤦

Reply:
Reduced the needed code to what the tests showed was the minimum required. Fixed via https://github.com/ampproject/amp-toolbox-php/pull/99/commits/4758e217229fb333a029cdc238c07cf2d459466c